### PR TITLE
Add unbound analysis support for create analyzer statement.

### PR DESCRIPTION
**Pull request by kovrus**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/CreateAnalyzerStatementAnalyzer.java [hunk 1]
```diff
@@ -21,240 +21,156 @@
...
-        for (NamedProperties tokenFilterNode : tokenFilters.tokenFilters()) {
+        for (AnalyzerElement<Expression> element : createAnalyzer.elements()) {
+            AnalyzerElementsAnalysisVisitor.analyze(element, context);
+        }
+        return analyzedCreateAnalyzer;
+    }
 
-            String name = tokenFilterNode.ident();
-            GenericProperties<Expression> properties = tokenFilterNode.properties();
+    private static class AnalyzerElementsAnalysisVisitor
```
Comment (kovrus) [inline sql/src/main/java/io/crate/analyze/CreateAnalyzerStatementAnalyzer.java]: I've introduced the `AnalyzerElementsAnalysisVisitor` only for the analyzer elements, before the `CreateAnalyzerStatementAnalyzer` was extending the visitor methods including the `visitCreateAnalyzer` which seems to be not that needed and a bit confusing.

File: sql/src/main/java/io/crate/analyze/AnalyzedCreateAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    void addTokenFilter(String name, GenericProperties<Symbol> properties) {
+        tokenFilters.put(name, properties);
+    }
+
+    void addGenericAnalyzerProperty(GenericProperty<Symbol> property) {
+        this.genericAnalyzerProperties.add(property);
+    }
+
+    void addCharFilter(String name, GenericProperties<Symbol> properties) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/AnalyzedCreateAnalyzer.java]: Could we make the analyzed statement class itself immutable ?

With prepared statements we can cache the analyzed statements and re-use them across query execution. Having this immutable would eliminate one whole class of bugs.

Comment (kovrus) [reply sql/src/main/java/io/crate/analyze/AnalyzedCreateAnalyzer.java]: yep, will do

File: sql/src/main/java/io/crate/analyze/CreateAnalyzerStatementAnalyzer.java [hunk 3]
```diff
@@ -21,240 +21,172 @@
...
     }
 
-    static class Context {
-
-        final Analysis analysis;
-        CreateAnalyzerAnalyzedStatement statement;
-
-        Context(Analysis analysis) {
-            this.analysis = analysis;
+    private static class Context {
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/CreateAnalyzerStatementAnalyzer.java]: I think it doesn't matter too much which way. I'd keep it as is now.

**Pull request merged**